aws/resource_aws_codeartifact_domain.go

```diff
@@ -33,7 +33,7 @@ func resourceAwsCodeArtifactDomain() *schema.Resource {
 			},
 			"encryption_key": {
 				Type:         schema.TypeString,
-				Required:     true,
+				Optional:     true,
```

Comment:
It is optional in the SDK: https://docs.amazonaws.cn/sdk-for-go/api/service/codeartifact/#CreateDomainInput
And in the API: https://docs.aws.amazon.com/codeartifact/latest/APIReference/API_CreateDomain.html#codeartifact-CreateDomain-request-encryptionKey

Reply:
CC @DrFaust92

Reply:
Hey @j3parker, 
I'll take a closer look but from what I recall the API validates that the kms Id field is not empty even if it's optional. That's why it's required in terraform.

Reply:
Here's what I get from the CLI:

```
$ aws configure set cli_history enabled

$ aws codeartifact create-domain --domain test
{
    "domain": {
        "name": "test",
        "owner": "(redacted)",
        "arn": "arn:aws:codeartifact:us-east-1:(redacted):domain/test",
        "status": "Active",
        "createdTime": "2020-10-09T10:33:21.559000-04:00",
        "encryptionKey": "arn:aws:kms:us-east-1:(redacted):key/(redacted)",
        "repositoryCount": 0,
        "assetSizeBytes": 0
    }
}

$ aws history show

AWS CLI command entered
at time: 2020-10-09 10:33:20.776
with AWS CLI version: aws-cli/2.0.49 Python/3.7.7 Windows/10 exe/AMD64
with arguments: ['codeartifact', 'create-domain', '--domain', 'test']

[0] API call made
at time: 2020-10-09 10:33:20.930
to service: codeartifact
using operation: CreateDomain
with parameters: {
    "domain": "test"
}

[0] HTTP request sent
at time: 2020-10-09 10:33:20.936
to URL: https://codeartifact.us-east-1.amazonaws.com/v1/domain?domain=test
with method: POST
with headers: {
    "Authorization": "(redacted)",
    "Content-Length": "0",
    "User-Agent": "aws-cli/2.0.49 Python/3.7.7 Windows/10 exe/AMD64 command/codeartifact.create-domain",
    "X-Amz-Date": "20201009T143320Z",
    "X-Amz-Security-Token": "(redacted)"
}
with body: There is no associated body

[0] HTTP response received
at time: 2020-10-09 10:33:21.466
with status code: 200
with headers: {
    "Connection": "keep-alive",
    "Content-Length": "354",
    "Content-Type": "application/json",
    "Date": "Fri, 09 Oct 2020 14:33:21 GMT",
    "x-amzn-RequestId": "(redacted)"
}
with body: {
    "domain": {
        "arn": "arn:aws:codeartifact:us-east-1:(redacted):domain/test",
        "assetSizeBytes": 0,
        "createdTime": 1602254001.559,
        "encryptionKey": "arn:aws:kms:us-east-1:(redacted):key/(redacted)",
        "name": "test",
        "owner": "(redacted)",
        "repositoryCount": 0,
        "s3BucketArn": "arn:aws:s3:::assets-(redacted)-us-east-1",
        "status": "Active"
    }
}
```

So it looks like it's truly optional (it's not needed even on the wire).
If it's not supplied it defaults to an AWS-managed CMK which is pretty typical for AWS services (they get created once per account, but only on-demand. AWS takes care of rotating them, they're free, you don't typically need to specify them, etc.)